### ISSUE TITLE
Empty BLAST output results in error message 

### ISSUE THREAD

**Issue by asczyrba** [widdowquinn/pyani#27]
I have a couple of hundred genomes I would like to calculate the ANI for using blastn.
Some pairs of genomes do not produce any alignments and therefore empty BLAST files.
This results in an error message during the parsing step and the pipeline just breaks.

Could you check for this case and fill the table with reasonable values (mainly zeros) so that 
the tables and graphics still get generated?

Thanks,
Alex

**Comment by widdowquinn** [comment]
Thanks for the heads-up, Alex. Do you have some example data you can point me to or share, so that I can confirm the issue and fix?

**Comment by asczyrba** [comment]
Here is example data. Thanks for your quick reply!
Alex

ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA_000494285.1_Cloacimonetes_bacterium_OTU_4/GCA_000494285.1_Cloacimonetes_bacterium_OTU_4_genomic.fna.gz
ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000008885.1_ASM888v1/GCF_000008885.1_ASM888v1_genomic.fna.gz
ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000069025.1_ASM6902v1/GCF_000069025.1_ASM6902v1_genomic.fna.gz
ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000283615.1_ASM28361v1/GCF_000283615.1_ASM28361v1_genomic.fna.gz

Traceback (most recent call last):
  File "/usr/local/bin/average_nucleotide_identity.py", line 772, in <module>
    results = methods[args.method][0](infiles, org_lengths)
  File "/usr/local/bin/average_nucleotide_identity.py", line 550, in unified_anib
    fraglengths=fraglengths, mode=args.method)
  File "/usr/local/lib/python3.5/dist-packages/pyani/anib.py", line 373, in process_blast
    mode)
  File "/usr/local/lib/python3.5/dist-packages/pyani/anib.py", line 425, in parse_blast_tab
    data = pd.DataFrame.from_csv(filename, header=None, sep='\t')
  File "/usr/local/lib/python3.5/dist-packages/pandas/core/frame.py", line 1189, in from_csv
    infer_datetime_format=infer_datetime_format)
  File "/usr/local/lib/python3.5/dist-packages/pandas/io/parsers.py", line 562, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/usr/local/lib/python3.5/dist-packages/pandas/io/parsers.py", line 315, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/usr/local/lib/python3.5/dist-packages/pandas/io/parsers.py", line 645, in __init__
    self._make_engine(self.engine)
  File "/usr/local/lib/python3.5/dist-packages/pandas/io/parsers.py", line 799, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/usr/local/lib/python3.5/dist-packages/pandas/io/parsers.py", line 1213, in __init__
    self._reader = _parser.TextReader(src, **kwds)
  File "pandas/parser.pyx", line 523, in pandas.parser.TextReader.__cinit__ (pandas/parser.c:5214)
pandas.io.common.EmptyDataError: No columns to parse from file

**Comment by widdowquinn** [comment]
That's great, thank you!

As a first comment, I thought I was handling this sort of thing already, but the error's being thrown in the `Pandas` import (which I moved to in v0.2) - this is a gap in my testing, as much as anything else. Should be pretty straightforward to fix, and I hope to have it done, soon.

**Comment by widdowquinn** [comment]
Just a note to say I see the same error using those sequences:

```
$ average_nucleotide_identity.py -i test_fail -o test_fail_ANIb -m ANIb -g --gmethod seaborn --gformat pdf,png -v 
[...]
INFO: All multiprocessing jobs complete.
INFO: Processing pairwise ANIb BLAST output.
Traceback (most recent call last):
  File "/Users/lpritc/Virtualenvs/pyani3/bin/average_nucleotide_identity.py", line 4, in <module>
    __import__('pkg_resources').run_script('pyani==0.2.0.dev0', 'average_nucleotide_identity.py')
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pkg_resources/__init__.py", line 719, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pkg_resources/__init__.py", line 1511, in run_script
    exec(script_code, namespace, namespace)
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pyani-0.2.0.dev0-py3.5.egg/EGG-INFO/scripts/average_nucleotide_identity.py", line 772, in <module>
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pyani-0.2.0.dev0-py3.5.egg/EGG-INFO/scripts/average_nucleotide_identity.py", line 550, in unified_anib
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pyani-0.2.0.dev0-py3.5.egg/pyani/anib.py", line 372, in process_blast
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pyani-0.2.0.dev0-py3.5.egg/pyani/anib.py", line 417, in parse_blast_tab
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pandas/core/frame.py", line 1189, in from_csv
    infer_datetime_format=infer_datetime_format)
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pandas/io/parsers.py", line 562, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pandas/io/parsers.py", line 315, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pandas/io/parsers.py", line 645, in __init__
    self._make_engine(self.engine)
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pandas/io/parsers.py", line 799, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/Users/lpritc/Virtualenvs/pyani3/lib/python3.5/site-packages/pandas/io/parsers.py", line 1213, in __init__
    self._reader = _parser.TextReader(src, **kwds)
  File "pandas/parser.pyx", line 523, in pandas.parser.TextReader.__cinit__ (pandas/parser.c:5214)
pandas.io.common.EmptyDataError: No columns to parse from file
```

**Comment by widdowquinn** [comment]
I think I have `pyani`'s `ANIb` mode working with your data, now - so I'm moving on to fixing `ANIm`, which currently hangs on the data. Once I could run the tool on your examples, there appeared to be very little in the way of homologous sequence between isolates (<=17kbp/1.5% of the genome for any pairwise comparison). On a scientific note, my opinion is that at this level of divergence ANI tells you very little, and may actively overemphasise a tiny, perhaps random, similarity between genomes - you probably already think this… ;)

![anib_alignment_coverage](https://cloud.githubusercontent.com/assets/63981/15932222/c0fd8b16-2e52-11e6-8036-1097da364988.png)
![anib_alignment_lengths](https://cloud.githubusercontent.com/assets/63981/15932219/c0ece338-2e52-11e6-91ee-b20d5ac5a4d6.png)
![anib_percentage_identity](https://cloud.githubusercontent.com/assets/63981/15932220/c0edcc8a-2e52-11e6-9ee9-01f8fa1c5104.png)
![anib_similarity_errors](https://cloud.githubusercontent.com/assets/63981/15932221/c0ef71ac-2e52-11e6-815d-cafea1a3b1cf.png)

**Comment by widdowquinn** [comment]
I think that, as of these commits, this bug is fixed.

Thank you again for raising the issue - this has helped greatly. If the problem persists I'll reopen this issue. If there are any other problems, please do get in touch!

**Comment by asczyrba** [comment]
Thank you very much!
I agree that an ANI for genomes with very little overlap is not very useful.
An option to filter these would be helpful (maybe a filter for minimal coverage).

**Comment by widdowquinn** [comment]
A relatively simple thing to do would be to take the Hadamard product of the percentage identity and coverage… I'll add that to the output. I don't know how meaningful it would be, yet…

**Comment by widdowquinn** [comment]
That's Hadamard product implemented in commit 093aee2e - I'd be interested to see how it goes with your data, if you'd like to try it.
